### PR TITLE
`Controlled` Bloq learns `add_my_tensors`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Qualtran/issues/650

Currently based on top of https://github.com/quantumlib/Qualtran/pull/809

### CODE REVIEW

File: qualtran/_infra/controlled.py [hunk 1]
```diff
@@ -318,6 +380,43 @@ def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT
...
+        in_ind = _flatten_soquet_collection(incoming[reg.name] for reg in self.signature.lefts())
+        out_ind = _flatten_soquet_collection(outgoing[reg.name] for reg in self.signature.rights())
+        assert data.shape == tuple(2**soq.reg.bitsize for ind in [out_ind, in_ind] for soq in ind)
+        # Figure out the ctrl indexes for which the ctrl is "active"
+        active_idx: List[Union[int, slice]] = [slice(x) for x in data.shape]
+        ctrl_idx = 0
+        for cv, ctrl_reg in zip(self.ctrl_spec.cvs, self.ctrl_regs):
+            for idx in ctrl_reg.all_idxs():
+                active_idx[ctrl_idx] = int(cv[idx])
+                active_idx[ctrl_idx + len(out_ind)] = int(cv[idx])
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:411]: what is going on here? what's the ` + len(out_ind)`? should `active_idx` be two dimensional?

File: qualtran/_infra/controlled.py [hunk 2]
```diff
@@ -318,6 +380,43 @@ def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT
...
+            np.eye(2 ** self.signature.n_qubits(), dtype=np.complex128), self.signature
+        )
+        # Verify it has the right shape
+        in_ind = _flatten_soquet_collection(incoming[reg.name] for reg in self.signature.lefts())
+        out_ind = _flatten_soquet_collection(outgoing[reg.name] for reg in self.signature.rights())
+        assert data.shape == tuple(2**soq.reg.bitsize for ind in [out_ind, in_ind] for soq in ind)
+        # Figure out the ctrl indexes for which the ctrl is "active"
+        active_idx: List[Union[int, slice]] = [slice(x) for x in data.shape]
+        ctrl_idx = 0
+        for cv, ctrl_reg in zip(self.ctrl_spec.cvs, self.ctrl_regs):
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:408]: if I understand correctly, this is highly specific to the current concrete implementation of `CtrlSpec`. You can imagine a world where there is an abstract `CtrlSpec` base class and multiple implementations that can support things other than conjunction of a bunch of equals-a-value. In my head, `cvs` would be an implementation detail and `CtrlSpec.is_active` would be the interface.

File: qualtran/cirq_interop/_cirq_to_bloq.py [hunk 3]
```diff
@@ -179,21 +178,18 @@ def _wire_symbol_from_gate(gate: cirq.Gate, signature: Signature, soq: 'Soquet')
...
+        [
+            *itertools.chain.from_iterable(
+                (2**reg.bitsize,) * int(np.prod(reg.shape))
+                for reg in [*signature.rights(), *signature.lefts()]
+            )
+        ]
+    )
+
+
+def _tensor_data_from_unitary_and_signature(inp_unitary: np.ndarray, signature: Signature):
```
Comment (mpharrigan) [inline qualtran/cirq_interop/_cirq_to_bloq.py:192]: move to a module in `qualtran.simulation.tensor`?

File: qualtran/_infra/controlled.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py:408]: At that point, we could iterate over all `2 ** n` possible bitstrings that `cvs` can take and call `is_active` to check if that subspace should be "Activated"

Comment (tanujkhattar) [reply qualtran/_infra/controlled.py:408]: Note `SumOfProducts` is a generalized representation that can be used to represent any arbitrary boolean function.

File: qualtran/_infra/controlled.py [hunk 4]
```diff
@@ -318,6 +380,43 @@ def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT
...
+            for idx in ctrl_reg.all_idxs():
+                active_idx[ctrl_idx] = int(cv[idx])
+                active_idx[ctrl_idx + len(out_ind)] = int(cv[idx])
+                ctrl_idx += 1
+        # Put the subbloq tensor at indices where ctrl is active.
+        subbloq_shape = _tensor_shape_from_signature(self.subbloq.signature)
+        data[tuple(active_idx)] = self.subbloq.tensor_contract().reshape(subbloq_shape)
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:415]: consider factoring out a lot of this into a function in `qualtran.simulation.tensor`. Perhaps: creating the `data` array; and then a function that uses it to mangle the indices

File: qualtran/_infra/controlled.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py:411]: We typically pass `out_idx + inp_idx` to the `tn.add()` call. The `active_idx`  is a tuple of length `len(out_idx.shape) + len(inp_idx.shape)`; which is equal to `len(data.shape)` and is used to find the subspace of `data` (an n-d array) that should be actived. 

Thinking of CSWAP;  if you want to express `data[1, :, :, 1, :, :] = _swap_matrix()`; `active_idx` will be a tuple corresponding to `(1, :, :, 1, :, :)`

File: qualtran/cirq_interop/_cirq_to_bloq.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/_cirq_to_bloq.py]: done

File: qualtran/_infra/controlled.py (hunk 4 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py]: done

**Comment by tanujkhattar** [comment]
@mpharrigan I've addressed all comments, PTAL

File: qualtran/simulation/tensor/_tensor_data_manipulation.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    out_shape, inp_shape = tensor_out_inp_shape_from_signature(signature)
+    return out_shape + inp_shape
+
+
+def active_space_for_ctrl_spec(
+    signature: Signature, ctrl_spec: CtrlSpec
+) -> Tuple[Union[int, slice], ...]:
+    """Returns the "active" subspace corresponding to `signature` and `ctrl_spec`.
+
+    Assumes first n-registers for `signature` are control registers corresponding to `ctrl_spec`.
```
Comment (mpharrigan) [inline qualtran/simulation/tensor/_tensor_data_manipulation.py:67]: say somewhere that this is a sequence of indices/slices that can address into a ndarray

**Review by mpharrigan — APPROVED** [review]

File: qualtran/simulation/tensor/_tensor_data_manipulation.py (hunk 5 above)
Comment (tanujkhattar) [reply qualtran/simulation/tensor/_tensor_data_manipulation.py:67]: done

**Pull request merged**